codec/prometheus/codecprometheus.go

```diff
@@ -0,0 +1,218 @@
+package codecprometheus
+
+import (
+	"context"
+	"fmt"
+	"regexp"
+	"strings"
+	"time"
+
+	"github.com/tsaikd/gogstash/config"
+	"github.com/tsaikd/gogstash/config/goglog"
+	"github.com/tsaikd/gogstash/config/logevent"
+)
+
+// ModuleName is the name used in config file
+const ModuleName = "prometheus"
+
+// ErrorTag tag added to event when process module failed
+const ErrorTag = "gogstash_codec_prometheus_error"
+
+// Codec default struct for codec
+type Codec struct {
+	config.CodecConfig
+}
+
+var (
+	fullStringMatch      = `(\S+)\{([^\}]+)\}\s+(.+)`
+	fullStringMatchRegex *regexp.Regexp
+)
+
+// InitHandler initialize the codec plugin
+func InitHandler(context.Context, config.ConfigRaw) (config.TypeCodecConfig, error) {
+	var err error
+	fullStringMatchRegex, err = regexp.Compile(fullStringMatch)
+	return &Codec{
+		CodecConfig: config.CodecConfig{
+			CommonConfig: config.CommonConfig{
+				Type: ModuleName,
+			},
+		},
+	}, err
+}
+
+// Decode returns an event from 'data' as JSON format, adding provided 'eventExtra'
+func (c *Codec) Decode(ctx context.Context, data any, eventExtra map[string]any, tags []string, msgChan chan<- logevent.LogEvent) (ok bool, err error) {
+	event := logevent.LogEvent{
+		Timestamp: time.Now(),
+		Extra:     eventExtra,
+	}
+	if eventExtra == nil {
+		event.Extra = map[string]any{}
+	}
+	event.AddTag(tags...)
+
+	datastr := ""
+	switch data := data.(type) {
+	case string:
+		datastr = data
+	case []byte:
+		datastr = string(data)
+	default:
+		err = config.ErrDecodeData
+		event.AddTag(ErrorTag)
+		goglog.Logger.Error(err)
+		return false, err
+	}
+
+	// Split into individual metrics
+
+	// Convert to string, make sure we have a \n suffix
+	if !strings.HasSuffix(datastr, "\n") {
+		datastr += "\n"
+	}
+
+	lines := strings.Split(datastr, "\n")
+	savetype := map[string]string{}
+	for _, line := range lines {
+		if len(strings.TrimSpace(line)) == 0 {
+			continue
+		}
+
+		// fmt.Printf("Decode line = '%s'\n", line)
+		// Preserve formatted type definitions
+		if strings.HasPrefix(line, "# TYPE") {
+			p := strings.Split(line, " ")
+			if len(p) < 4 {
+				goglog.Logger.Error(fmt.Errorf("prometheus: TYPE line: %s", line))
+				continue
+			}
+			savetype[p[2]] = p[3]
+		}
+		// Skip line if comment, we don't process them further
+		if strings.HasPrefix(line, "#") {
+			continue
+		}
+
+		// Decode lines, every metric is an possible individual event
+		{
+			e := event
+			err = c.decodePrometheusEvent(line, savetype, &e)
+			if err == nil {
+				msgChan <- e
+				ok = true
+			}
```

Comment:
ignore decodePrometheusEvent error?

Reply:
That only pushes events into the channel if there are no errors in decoding.